README.md
# Completeness

"Completeness" calculates percentage of populated fields of the object.

It helps answering common questions: 'how complete the user profile is?' and 'what information is missing?'.


## Installation

Add to Gemfile

    gem git: 'git@github.com:DimaSamodurov/completeness.git'

## Usage

```ruby
class Profile
  attr_accessor :first_name, :last_name, :email, :phone, :addresses

  include Completeness

  define_completeness(
    {
        first_name:  { :if => 'present?', add: 20 },
        last_name:   { :if => 'present?', add: 20 },
        email:       { :if => 'present?', add: 40, boolean_method: 'email_provided?' },
        addresses:   { :if => 'any?',     add: 20, boolean_method: 'address_provided?' },
    }
  )
end

p = Profile.new
p.completeness         # => 0
p.complete?            # => false
p.first_name, p.last_name, p.email = %w(John Doe john.doe@mail.net)
p.completeness         # => 80
p.complete?            # => false
p.address_provided?    # => false
p.addresses = ['Crater Grimaldi, Moon, 5.2°S 68.6°W']
p.address_provided?    # => true
p.completeness         # => 100
p.complete?            # => true
```


## Testing

    rake test

## Doc

    yard doc

## Possible enhancement

- allow proc to be passed as "if" option
- validate sum = 100%
- eliminate ActiveSupport depencency
- configure defaults like {:if => 'nil?'}



HISTORY.md
# 0.0.4 - 30 Aug, 2012

## Features

- Boolean methods: user.email_provided? can be defined automatically based on completeness definition.
- New initialization method 'define_completeness'.
  Assigning completeness_shares attribute is deprecated and will be removed in version 0.1.0.
- Completeness definition validattion: sum of weights must equal 100 by default.


test/minitest_helper.rb
require 'minitest/spec'
require 'minitest/autorun'
require_relative '../lib/completeness'


test/completeness_test.rb
require_relative 'minitest_helper'

class Profile

  attr_accessor :first_name, :last_name, :email, :phone, :addresses

  include Completeness

  self.define_completeness(
      {
          first_name:  { :if => 'present?', weight: 20 },
          last_name:   {                    weight: 20 }, # used default :if
          email:       {                    weight: 30 },
          phone:       {                    weight: 10, boolean_method: 'phone_provided?' },
          addresses:   { :if => 'any?',     weight: 20, boolean_method: 'address_provided?' },
      }
  )
end

describe Completeness do

  describe "completeness_shares" do
    it "define completeness percent of fields" do
      p = Profile.new
      p.completeness.must_equal 0
      p.complete?.must_equal false

      p.first_name = 'John' and p.completeness.must_equal 20
      p.last_name = 'Doe' and p.completeness.must_equal 40
      p.email = 'john.doe@mail.net' and p.completeness.must_equal 70
      p.phone = '333222' and p.completeness.must_equal 80
      p.addresses = ["City,State,Zip"] and p.completeness.must_equal 100

      p.complete?.must_equal true
    end

    describe " 'if' condition" do
      it "default is 'present?' " do
        p = Profile.new
        p.email = '   '
        p.completeness_of(:email).must_equal 0 # as spaces considered as blank.
        p.email = 'john@mail.com'
        p.completeness_of(:email).must_equal 30
      end
    end

    it "can be overridden at the object level" do
      p = Profile.new
      p.completeness_shares =
          {
            first_name:  { weight: 45.5 },
            email:       { weight: 54.5 }
          }
      p.first_name = 'Joe'
      p.completeness.must_equal 45.5
      p.email = 'joe@mail.net'
      p.completeness.must_equal 100
      p.complete?.must_equal true
    end
  end

  describe "completeness_of(field)" do
    it "should return :weight of the field specified if the field conform 'if' condition" do
      p = Profile.new
      p.first_name = 'John'
      p.completeness_of(:first_name).must_equal 20
    end

    it "should raise exception for unknown field" do
      p = Profile.new
      proc { p.completeness_of(:abrakadabra) }.must_raise RuntimeError
    end
  end

  describe "completeness_title_of(field)" do
    it "should return :title option if specified" do
      p = Profile.new
      p.completeness_shares = { first_name:  { weight: 45.5, title: "Your name" } }
      p.completeness_title_of(:first_name).must_equal "Your name"
    end

    it "should return humanized field name if :title option is not specified" do
      p = Profile.new
      p.completeness_shares = { first_name:  { weight: 45.5 } }
      p.completeness_title_of(:first_name).must_equal "First Name"
    end
  end

  it "can be included to the object" do
    class Foo
      attr_accessor :baz, :bar
    end
    f = Foo.new
    f.singleton_class.instance_eval do
      include Completeness
      self.completeness_shares = { baz: { weight: 41 }, bar: { weight: 59 } }
    end
    f.complete?.must_equal false
    f.baz = 'catch'
    f.completeness.must_equal 41
    f.bar = 'me'
    f.completeness.must_equal 100
  end

  describe "incomplete_items" do
    it "should return names of items having zero completeness" do
      p = Profile.new
      p.incomplete_fields.must_equal [:first_name, :last_name, :email, :phone, :addresses]
      p.first_name, p.last_name = %w(John Doe)
      p.incomplete_fields.must_equal [:email, :phone, :addresses]
      p.email, p.phone, p.addresses = ['john@mail.net', '333222', ['Location1']]
      p.incomplete_fields.must_equal []
    end
  end

  describe "boolean_method" do
    it "defines instance method with name specified" do
      klass = Class.new do
        attr_accessor :email, :phone

        include Completeness
        define_completeness( {
                                email: { weight: 40 },
                                phone: { weight: 60, boolean_method: 'phone_provided?' }
                             }
        )
      end
      profile = klass.new
      profile.phone_provided?.must_equal false
      profile.phone = '333-222'
      profile.phone_provided?.must_equal true
    end
  end

  describe "completeness_validate_shares" do
    before do
      @klass = Class.new do
        attr_accessor :email, :phone

        include Completeness
      end
    end

    it "raise exception if sum of weights <> 100%" do
      proc {
        @klass.class_eval do
          define_completeness( { email: { weight: 0 }, phone: { weight: 90 } } )
        end
      }.must_raise RuntimeError
    end

    it "does not raise exception if sum of weights == 100%" do
      @klass.class_eval do
        define_completeness( { email: { weight: 10 }, phone: { weight: 90 } } )
      end
    end

  end
end


lib/completeness.rb
require "completeness/version"
require "active_support/concern"
require "active_support/core_ext/object/blank"
require "active_support/core_ext/object/try"
require "active_support/core_ext/class"


# Helps calculate the percent of populated fields of the object
# ==== Examples
#   class Profile
#     attr_accessor :addresses
#     include Completeness
#     completeness_shares = { addresses: {:if => 'any?', weight: 20} }
#   end
#
#   p = Profile.new
#   p.completeness # => 0 . No exception even 'addresses' is nil and does not respond to 'any?'
#   p.addresses = []
#   p.completeness # => 0. Still works as [].any? == false
#   p.addresses << 'Location 1'
#   p.completeness # => 20
#
module Completeness
  extend ActiveSupport::Concern

  #completeness_shares =
  #    {
  #        contact_email:  { :if => 'present?', weight: 60 },
  #        contact_phone:  { :if => 'present?', weight: 40 },
  #    }

  included do
    self.class_attribute :completeness_defaults
    self.completeness_defaults = {:if => "present?"}

    self.class_attribute :completeness_shares
    self.completeness_shares = {}
  end

  module ClassMethods

    def define_completeness(shares, custom_options = {} )
      default_options = {
          define_boolean_methods: true,
          validate_shares: true
      }
      options = default_options.merge custom_options
      self.completeness_shares = shares

      completeness_validate_shares        if options[:validate_shares]
      completeness_define_boolean_methods if options[:define_boolean_methods]
    end

    # Returns completeness weight option of the field
    def completeness_weight_of(field)
      completeness_share_of(field)[:weight]
    end

    # Returns title of the field:
    # - as :title option
    # - as humanized field name if :title option is not specified
    def completeness_title_of(field)
      completeness_share_of(field)[:title] || field.to_s.humanize.split.map(&:capitalize).join(' ')
    end

    # Return completeness weight of the field configured for current object
    def completeness_share_of(field)
      completeness_shares[field] or raise "Completeness share is not defined for '#{field}' of #{self.class.name}"
    end

    protected

    # let we have defined completeness on user model like:
    #  {
    #    email:      { title: 'Email address', weight: 80 }
    #    addresses:  { title: 'Addresses',  weight: 20, boolean_method: 'address_provided' },
    #  }
    # then you will get user.address_provided? method defined which return true in case weight of this field > 0.
    # we defined additional field instead of using user.addresses.present? because it will return true for empty array.
    def completeness_define_boolean_methods
      completeness_shares.each do |field, meta|
        if meta[:boolean_method]

          method_name = "#{meta[:boolean_method]}".to_sym

          raise "Oops, #{method_name} is already defined on the #{self.class.name}." +
                " Consider another name for boolean field." if method_defined? method_name

          define_method method_name do
            completeness_of(field) > 0
          end
        end
      end
    end

    def completeness_validate_shares
      sum = completeness_shares.values.map{|share| share[:weight]}.inject(0) { |sum, value| sum + value }
      raise "Expected completeness weights sum to be 100, but got #{sum}." if sum != 100
    end
  end

  include ClassMethods

  # More complex condition can be if object has several states depended on e.g workflow or permissions.
  # In that case list of shares can vary for different objects.
  # Then we can add flipper like :when => lambda { self.has_extended_properties? }


  # Determines completeness of 'field' based on +completeness_shares+ conditions.
  # Conditions are verified softly, not raising exception when rule is not applicable.
  def completeness_of(field)
    share = completeness_share_of(field)
    self.send(field).try((share[:if] || self.completeness_defaults[:if]).to_sym) ? share[:weight] : 0
  end

  # Calculates sum of weights of all complete fields
  def completeness
    completeness_shares.keys.inject(0) do |result, field|
      result += completeness_of(field)
    end
  end

  # True if completeness = 100%. In other words if all fields are complete.
  def complete?
    completeness == 100
  end

  # Returns fields whose completeness = 0
  def incomplete_fields
    completeness_shares.keys.select{|field| completeness_of(field) == 0 }
  end

end